alacritty_terminal/src/input.rs

```diff
@@ -198,10 +198,21 @@ impl<T: Eq> Binding<T> {
 
     #[inline]
     pub fn triggers_match(&self, binding: &Binding<T>) -> bool {
-        self.trigger == binding.trigger
-            && self.mods == binding.mods
-            && (self.mode.contains(binding.mode) || binding.mode.contains(self.mode))
-            && (self.notmode.contains(binding.notmode) || binding.notmode.contains(self.notmode))
+        // Check the binding's key and modifiers
+        if self.trigger != binding.trigger || self.mods != binding.mods {
+            return false;
+        }
+
+        // Completely empty modes match all modes
+        if (self.mode.is_empty() && self.notmode.is_empty())
+            || (binding.mode.is_empty() && binding.notmode.is_empty())
+        {
+            return true;
+        }
+
+        // Check for intersection (equality is required since empty does not intersect itself)
+        (self.mode == binding.mode || self.mode.intersects(binding.mode))
+            && (self.notmode == binding.notmode || self.notmode.intersects(binding.notmode))
```

Comment:
Depending on how you feel, we might be able to make this a bit clearer by breaking this into a new helper function: `modes_match`, which would deal with both of these last two chained conditionals. Just might be a little nicer since the first conditional is just about the keyboard input, while the second two are about the terminal's mode.

Reply:
Problem is that this is just something we want to use for this specific function. In the other areas where we match the different modes we do things differently, which is why I haven't done this yet.

It's definitely a bit confusing with how contains and intersects work, but I feel like this might be better than a bigger separation.